include/alpaka/mem/buf/Traits.hpp

```diff
@@ -69,6 +77,25 @@ namespace alpaka
         return traits::BufAlloc<TElem, Dim<TExtent>, TIdx, TDev>::allocBuf(dev, extent);
     }
 
+    //! The stream-ordered memory buffer type trait alias template to remove the ::type.
+    template<typename TDev, typename TElem, typename TDim, typename TIdx>
+    using AsyncBuf = typename traits::AsyncBufType<alpaka::Dev<TDev>, TElem, TDim, TIdx>::type;
```

Comment:
I'm not convinced that this is needed with this approach, since the `AsyncBufType` should always be the same as the `BufType` when it is defined.
On the other hand, since currently this is only defined for the CPU and CUDA devices, it could be used to check if the async buffer functionality is actually implemented for a device ?

Reply:
So far it seems we don't need it. I would remove it.

Reply:
OK.

Reply:
Maybe add a default implementation for `AsyncBufAlloc` that fails and informs the user that the given buffer does not support stream-ordered memory allocation ?

Reply:
Or better to leave it undefined, so it can be checked with SFINAE ?

Reply:
Wouldn't that eliminate the purpose of alpaka, to give you a best effort implementation of a particular behavior across all accelerators? Your code would no longer be portable. And we can workaround the limitations on all backends by emplacing a host task into the queue, that frees the buffer.

In case you could query though, whether the underlying backend supports async alloc/free, what would you do with the information?

Reply:
> And we can workaround the limitations on all backends by emplacing a host task into the queue, that frees the buffer.

I agree with you.
Sorry, I didn't mean that one should _not_ implement the stream-ordered behaviour all accelerator types.
But this PR currently does not, and I thought it would be useful so make it explicit somehow.

I'm happy to remove the `AsyncBuf` trait and _not_ add any checks, with the goal to eventually support stream-ordered allocations on every accelerator.

> In case you could query though, whether the underlying backend supports async alloc/free, what would you do with the information?

I guess the user code could fall back to some private implementation, or adapt the algorithms to use the standard allocations with some extra synchronisations.

Reply:
IIUC only the HIP, OpenAcc and OpenMP5 backends are missing? Would it be too hard to provide a quick implementation for them? With a callback, similar to the CPU implementation? If it is, I could open an issue and implement it later.

Reply:
> I guess the user code could fall back to some private implementation, or adapt the algorithms to use the standard allocations with some extra synchronisations.

Hmm. I asked the question because I wondered if the user would then resort to anyway implement the async alloc/free themselves by placing a callback into the queue. Phrased differently: can the user actually implement something better than our fallback implementation of placing the callback. Maybe we should provide a query than. But we can also wait until someone actually needs this feature.

Reply:
I don't have experience with OpenACC and OpenMP5, so I'm not sure about those :-(

For HIP I don't think there is a CPU thread managed by Alpaka associated with the Queue, so we cannot use the same approach used for the PU buffer.
And, assuming it has the same limitations as CUDA that a callback cannot call a CUDA/HIP function, a callback cannot be used either.
The other implementation I can think of is a full-fledged caching memory allocator; that's my long term goal anyway, but not something I can implement quickly. Also, so far it's not clear if it would be desirable to have it inside Alpaka or built on top or on the side.

Reply:
> I don't have experience with OpenACC and OpenMP5, so I'm not sure about those :-(

Let's open an issue then when this PR is merged. @jkelling is experienced with those backends, so he might have a look later.

> And, assuming it has the same limitations as CUDA that a callback cannot call a CUDA/HIP function, a callback cannot be used either.

I think we have a workaround for that by keeping a background thread around. But I guess we can also implemented that later.